src/python/review/reviewers/utils/penalty.py

```diff
@@ -0,0 +1,137 @@
+    """
+    if lang_to_history is None:
+        return []
+
+    language_to_history = json.loads(lang_to_history)
+    history = language_to_history[language.value.lower()]
+
+    previous_issues = []
+    for issue_data in history:
+        previous_issues.append(PreviousIssue(**issue_data))
+    return previous_issues
+
+
+def categorize(previous_issues: List[PreviousIssue], current_issues: List[BaseIssue]):
+    """
+    For each previously made issue determines its category, with the help of current issues.
+    """
+    origin_class_to_category = {}
+    for issue in current_issues:
+        origin_class_to_category[issue.origin_class] = issue.type
+
+    for issue in previous_issues:
+        issue.category = origin_class_to_category.get(issue.origin_class, None)
+
+
+def get_issue_class_to_penalty(issues: List[PreviousIssue]) -> Dict[str, int]:
+    """
+    For each issue, the corresponding penalty coefficient is calculated.
+    """
+    return {
+        issue.origin_class: ISSUE_TYPE_TO_PENALTY_COEFFICIENT.get(issue.category, 1) * issue.quantity
+        for issue in issues
+    }
+
+
+def get_penalizing_classes(current_issues: List[BaseIssue], previous_issues: List[PreviousIssue]) -> Set[str]:
+    """
+    Returns issues that should be penalized.
+    """
+    current_classes = set(map(lambda issue: issue.origin_class, current_issues))
+    previous_classes = set(map(lambda issue: issue.origin_class, previous_issues))
+
+    return previous_classes.intersection(current_classes)
+
+
+def get_penalty_coefficient(current_issues: List[BaseIssue], previous_issues: List[PreviousIssue]) -> int:
+    """
+    Calculates the final penalty coefficient.
+    """
+
+    penalizing_classes = get_penalizing_classes(current_issues, previous_issues)
+    penalizing_issues = list(filter(lambda issue: issue.origin_class in penalizing_classes, previous_issues))
+
+    penalty_coefficient = 0
+    for issue in penalizing_issues:
+        penalty_coefficient += issue.quantity * ISSUE_TYPE_TO_PENALTY_COEFFICIENT[issue.category]
+
+    return penalty_coefficient
+
+
+def get_issue_influence_on_penalty(issue_penalty_coefficient: int, total_penalty_coefficient: int) -> int:
+    """
+    Calculates the influence of the issue on the penalty.
+
+    Returns a number in the range from 0 to 100.
+    """
+
+    return int(issue_penalty_coefficient / total_penalty_coefficient * 100)
+
+
+def get_penalty_score(penalty_coefficient: int) -> int:
+    """
+    Calculates the penalty score with the penalty coefficient
+
+    Returns a number equal to 0, 1, 2 or 3, which describes how many levels the grade should be lowered.
+    """
+
+    if penalty_coefficient < common_penalty_rule.one_level_penalty:
+        return 0
+    elif penalty_coefficient < common_penalty_rule.two_level_penalty:
+        return 1
+    elif penalty_coefficient < common_penalty_rule.three_level_penalty:
+        return 2
+    else:
+        return 3
```

Comment:
Please use the default value in the end (without else branch), if you forget to consider a case you will return None, it is incorrect behaviour.